internal/builders/docker/main.go

```diff
@@ -0,0 +1,79 @@
+// Copyright 2022 SLSA Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"errors"
+	"fmt"
+	"os"
+)
+
+func main() {
+	buildConfigPath := flag.String("build_config_path", "",
+		"Required - Path to a toml file containing the build configs.")
+	sourceRepo := flag.String("source_repo", "",
+		"Required - URL of the source repo.")
+	gitCommitHash := flag.String("git_commit_hash", "",
+		"Required - SHA1 Git commit digest of the revisoin for the soruce code to build the artefact from.")
+	builderImage := flag.String("builder_image", "",
+		"Required - URL indicating the Docker builder image, including a URI and image digest.")
+
+	flag.Parse()
+
+	// TODO(#1191): Do we need any verification steps?
+	// 1. sourceRepo is a valid URI
+	// 2. gitCommitHash is valid (we expect to see sha1: at the beginning of the input string)
+	// 3. Docker image is fully specified name@alg:digest
+	// 4. buildConfigPath is relative
+
+	artifacts := make(map[string]int)
```

Comment:
maybe you could keep the main command here mostly empty for now (besides the flag definitions), and move the flag to config code into a package `internal/builders/docker/config.go`:

```
type DockerBuildConfig struct {
  SourceRepo string
  SourceDigest string
  BuilderImage string
  BuildConfig 
}

func NewDockerBuildConfig(sourceRepo, sourcedigest, builderimage, buildconfig string) (*DockerBuildConfig, error) error {
  // perform validation and assemble the configuration
}

```

and the build code into `internal/builders/docker/build.go` with some code like 

```
type DockerBuild struct {
  BuildDefinition *slsa.BuildDefinition
  // other stuff to be added (like the docker compiler)
}

func NewDockerBuild(config *DockerBuildConfig) (*DockerBuild, error) {
  // takes the valid DockerBuildConfig and assembles the BuildDefinition
}

func (d *DockerBuild) BuildArtifact() ({}interface, error) {
  // Perform the build, return the subjects as you noted above
}

func (d *DockerBuild) GeneratePredicate() ({}interface, error){
  // "Dry-run" to generate the SLSA predicate
}
```

Reply:
Thanks. Yes, I will refactor this code :)

Reply:
I am not sure about 

```
func (d *DockerBuild) GeneratePredicate() ({}interface, error){
  // "Dry-run" to generate the SLSA predicate
}
```

At this stage we don't have all the information to generate a complete predicate. My plan for the dry is to just generate a `BuildDefinition`.

Reply:
> At this stage we don't have all the information to generate a complete predicate. My plan for the dry is to just generate a BuildDefinition.

I see -- you're right. That sounds good. I think this method can just go away then.

> If we want to return the full subject we need a convention for setting the subject name, for instance that the base name from the artifact_path will be the subject name. Does that convention make sense?

Right! Agree that the base of artifact_path can be used to start (and that works for multiple artifacts -- let's say that output/** is the artifact_path and there are output/foo and output/bar in there -- foo and bar can be used.) If we want to configure the subjects dir name I think we can add an option, but at least the base will stay fixed (I can't imagine why it wouldn't)